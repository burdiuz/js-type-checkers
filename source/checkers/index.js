import PrimitiveTypeChecker from './primitive';

let defaultTypeChecker = PrimitiveTypeChecker;

export const getDefaultTypeChecker = () => defaultTypeChecker;
export const setDefaultTypeChecker = (typeChecker) => {
  defaultTypeChecker = typeChecker;
};
