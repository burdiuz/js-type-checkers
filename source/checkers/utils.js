
export const GET_PROPERTY = '(GetProperty)';
export const SET_PROPERTY = '(SetProperty)';
export const INDEX = '(Index)';
export const ARGUMENTS = '(Arguments)';
export const RETURN_VALUE = '(ReturnValue)';
export const MERGE = '(Merge)';

export function AsIs(value) {
  if (this instanceof AsIs) {
    this.value = value;
  } else {
    return new AsIs(value);
  }
}

function asIs() {
  return this.value;
}

AsIs.prototype.toString = asIs;
AsIs.prototype.valueOf = asIs;
AsIs.prototype[Symbol.toPrimitive] = asIs;

export const buildPath = (sequence) => sequence
  .reduce((str, name) => {
    if (name instanceof AsIs) {
      str = `${str}${name}`;
    } else if (String(parseInt(name, 10)) === name) {
      str = `${str}[${name}]`;
    } else if (/^[a-z][\w$]*$/i.test(name)) {
      str = str ? `${str}.${name}` : name;
    } else {
      str = `${str}["${name}"]`;
    }

    return str;
  }, '');
