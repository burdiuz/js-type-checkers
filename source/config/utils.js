import hasOwn from '@actualwave/has-own';

export const singleConfigFactory = (defaultValue = null, validator = undefined) => {
  let value = defaultValue;

  return {
    get: () => value,
    set: (newValue = defaultValue) => {
      if (validator) {
        value = validator(newValue);
      } else {
        value = newValue;
      }
    },
  };
};

export const mapConfigFactory = (defaultValues = {}) => {
  const getDefault = () => ({ ...defaultValues });

  const values = getDefault();

  return {
    values,
    getDefault,
    set: (newValues) => Object.assign(values, newValues),
    get: () => ({ ...values }),
    getValue: (key, target = values) => (hasOwn(target, key) ? target[key] : undefined),
  };
};
