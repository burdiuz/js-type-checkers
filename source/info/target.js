import { createPathSequence } from '@actualwave/path-sequence-to-string';
import { createChildrenCache } from './children';

export const INFO_KEY = Symbol('type-checkers::info');

export const getTargetInfo = (target) => (target ? target[INFO_KEY] : undefined);

export const setTargetInfo = (target, info) => {
  if (target && info) {
    target[INFO_KEY] = info;
  }
};

export const removeTargetInfo = (target) => delete target[INFO_KEY];

export const hasTargetInfo = (target) => !!getTargetInfo(target);

class TargetInfo {
  constructor(
    checker,
    data = null,
    deep = true,
    names = createPathSequence(),
    children = createChildrenCache(),
  ) {
    this.checker = checker;
    this.data = data;
    this.deep = deep;
    this.names = names;
    this.children = children;
  }

  getChild(name) {
    return this.children.get(name);
  }

  storeChildFrom(name, child) {
    const info = getTargetInfo(child);

    if (info) {
      this.children.store(name, info);
    }
  }

  createChild(name, data = null) {
    const childInfo = new TargetInfo(this.checker, data, this.deep, this.names.clone(name));

    this.children.store(name, childInfo);

    return childInfo;
  }


  createChildWithNames(names, data = null) {
    const childInfo = new TargetInfo(this.checker, data, this.deep, names);

    this.children.store(names.lastName, childInfo);

    return childInfo;
  }

  copy({ deep, checker, children, data, names }) {
    if (this.checker === checker) {
      this.deep = this.deep || deep;
      this.children.copy(children);
      this.data = checker.mergeConfigs(this.data, data, names);
    } else {
      console.error(
        'TypeChecked objects can be merged only if using exactly same instance of type checker.',
      );
    }

    return this;
  }
}

export const createTargetInfo = (checker, data, deep, names, children) =>
  new TargetInfo(checker, data, deep, names, children);
