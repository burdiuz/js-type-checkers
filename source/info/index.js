import {
  INFO_KEY,
  getTargetInfo,
  setTargetInfo,
  removeTargetInfo,
  hasTargetInfo,
  createTargetInfo,
} from './target';

import { getTypeChecker, getTypeCheckerData } from './utils';

export {
  INFO_KEY,
  getTargetInfo,
  setTargetInfo,
  removeTargetInfo,
  hasTargetInfo,
  createTargetInfo,
  getTypeChecker,
  getTypeCheckerData,
};
