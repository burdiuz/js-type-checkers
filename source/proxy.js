import { getDefaultTypeChecker } from './defaultTypeChecker';
import { getErrorReporter } from './errorReporter';
import { isEnabled } from './enabled';
import { getConfig, setConfig, configKey } from './config';

const validTypes = {
  object: true,
  function: true,
};

export const isValidTarget = (target) => target && validTypes[typeof target];

const getProperty = (target, property) => {
  // if object or function and deep -- wrap
  const value = target[property];

  if (property === configKey) {
    return value;
  }

  const { deep, names, config, typeChecker } = getConfig(target);

  typeChecker.getProperty && typeChecker.getProperty(target, property, value, config, names);

  if ((deep && isValidTarget(value)) || value instanceof Function) {
    const { children } = getConfig(target);

    if (!children.hasOwnProperty(property)) {
      children[property] = create(value, { deep, names: [...names, property] }, typeChecker);
    }

    return children[property];
  }

  return value;
};

const setProperty = (target, property, value, receiver) => {
  const { names, config, children, typeChecker } = getConfig(target);

  if (property !== configKey) {
    delete children[property];
    typeChecker.setProperty && typeChecker.setProperty(target, property, value, config, names);
  }

  target[property] = value;
};

const callFunction = (target, thisArg, argumentsList) => {
  const { names, config, typeChecker } = getConfig(target);

  typeChecker.arguments && typeChecker.arguments(target, thisArg, argumentsList, config, names);

  const result = target.apply(thisArg, argumentsList);

  typeChecker.returnValue && typeChecker.returnValue(target, thisArg, result, config, names);

  return result;
};

const objectProxy = (target) => new Proxy(
  target,
  {
    get: getProperty,
    set: setProperty,
  },
);

const functionProxy = (target) => new Proxy(
  target,
  {
    apply: callFunction,
    construct: callFunction,
  },
);

export const create = (target, { deep = true, names = [] } = {}, typeChecker = getDefaultTypeChecker()) => {
  if (!isValidTarget(target) || !isEnabled() || getConfig(target)) {
    return target;
  }

  setConfig(
    target,
    {
      deep,
      names,
      children: {},
      typeChecker,
      config: typeChecker.init(target, getErrorReporter()),
    },
  );

  if (target instanceof Function) {
    return functionProxy(target);
  }

  return objectProxy(target);
};
