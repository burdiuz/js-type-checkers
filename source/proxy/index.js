import { getProxyConfig, setProxyConfig } from './config';
import create from './create';
import createDeep from './createDeep';

export {
  getProxyConfig,
  setProxyConfig,
  create,
  createDeep,
};
