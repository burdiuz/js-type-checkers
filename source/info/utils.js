import { INFO_KEY } from './target';

export const getTypeChecker = (target) => {
  if (target) {
    const info = target[INFO_KEY];

    return (info && info.checker) || undefined;
  }

  return undefined;
};

export const getTypeCheckerData = (target) => {
  if (target) {
    const info = target[INFO_KEY];

    return (info && info.data) || undefined;
  }

  return undefined;
};
