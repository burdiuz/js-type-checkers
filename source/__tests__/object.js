import { setErrorReporter } from '../reporters';
import { PrimitiveTypeChecker, setDefaultTypeChecker } from '../checkers';
import { SET_PROPERTY, GET_PROPERTY, INDEX, ARGUMENTS, RETURN_VALUE } from '../checkers/utils';
import { create, isEnabled, isTypeChecked, setEnabled, getTargetInfo, getTargetTypeCheckerConfig, setTargetInfo } from '../';
import { TARGET_KEY, getOriginalTarget } from '../target/proxy';
import { INFO_KEY } from '../target/info';

describe('Object', () => {
  const reporter = jest.fn();
  let target;

  beforeEach(() => {
    setErrorReporter(reporter);
    reporter.mockClear();
  });

  describe('When types pre-initialized', () => {
    beforeEach(() => {
      PrimitiveTypeChecker.collectTypesOnInit = true;
      target = create({
        numberValue: 12,
        stringValue: 'my string',
        booleanValue: true,
        arrayValue: [1, 2, 3],
        objectValue: { val1: 1, val2: '2', val3: true },
        method: (a) => !!a,
      });
    });

    describe('When getting values', () => {
      beforeEach(() => {
        expect(target.numberValue).toBe(12);
        expect(target.stringValue).toBe(12);
        expect(target.booleanValue).toBe(12);
      });
    });

    describe('When setting values of mismatched types', () => {
      beforeEach(() => {
        target.numberValue = '123';
        target.stringValue = false;
        target.booleanValue = false;
        target.arrayValue = {};
        target.objectValue = [];
      });

      it('should report mismatched primitive types', () => {
        expect(reporter).toHaveBeenCalledTimes(4);
        expect(reporter).toHaveBeenCalledWith(SET_PROPERTY, 'numberValue', 'number', 'string');
        expect(reporter).toHaveBeenCalledWith(SET_PROPERTY, 'stringValue', 'string', 'boolean');
        expect(reporter).toHaveBeenCalledWith(SET_PROPERTY, 'arrayValue', 'array', 'object');
        expect(reporter).toHaveBeenCalledWith(SET_PROPERTY, 'objectValue', 'object', 'array');
      });
    });

    describe('When calling method', () => {
      beforeEach(() => {
        expect(target.method(1)).toBe(true);
      });

      it('should not report any errors', () => {
        expect(reporter).not.toHaveBeenCalled();
      });

      describe('When calling method with same argument types', () => {
        beforeEach(() => {
          expect(target.method(0)).toBe(false);
        });

        it('should not report any errors', () => {
          expect(reporter).not.toHaveBeenCalled();
        });
      });

      describe('When calling method with mismatched argument types', () => {
        beforeEach(() => {
          expect(target.method('0')).toBe(true);
        });

        it('should report call errors', () => {
          expect(reporter).toHaveBeenCalledTimes(1);
          expect(reporter).toHaveBeenCalledWith(ARGUMENTS, 'method[0]', 'number', 'string');
        });
      });

      describe('When method is replaced in runtime', () => {
        beforeEach(() => {
          target.method = (b) => ++b;
        });

        describe('When calling method with mismatched argument types', () => {
          beforeEach(() => {
            expect(target.method('5')).toBe(6);
          });

          it('should report call errors', () => {
            expect(reporter).toHaveBeenCalledTimes(2);
            expect(reporter).toHaveBeenCalledWith(ARGUMENTS, 'method[0]', 'number', 'string');
            expect(reporter).toHaveBeenCalledWith(RETURN_VALUE, `method${RETURN_VALUE}`, 'boolean', 'number');
          });
        });
      });
    });
  });

  describe('When types were not initialized', () => {
    beforeEach(() => {
      PrimitiveTypeChecker.collectTypesOnInit = false;
      target = create({
        numberValue: 12,
        booleanValue: true,
        objectValue: {},
      });
    });

    describe('When setting values of mismatched types', () => {
      beforeEach(() => {
        target.numberValue = '123';
        target.stringValue = false;
        target.booleanValue = false;
        target.arrayValue = {};
        target.objectValue = [];
      });

      it('should record set types without error', () => {
        expect(reporter).not.toHaveBeenCalled();
      });

      describe('When setting values of mismatched types', () => {
        beforeEach(() => {
          target.numberValue = 123;
          target.stringValue = 'true';
          target.booleanValue = true;
          target.arrayValue = [];
          target.objectValue = {};
        });

        it('should record set types without error', () => {
          expect(reporter).toHaveBeenCalledTimes(4);
          expect(reporter).toHaveBeenCalledWith(SET_PROPERTY, 'numberValue', 'string', 'number');
          expect(reporter).toHaveBeenCalledWith(SET_PROPERTY, 'stringValue', 'boolean', 'string');
          expect(reporter).toHaveBeenCalledWith(SET_PROPERTY, 'arrayValue', 'object', 'array');
          expect(reporter).toHaveBeenCalledWith(SET_PROPERTY, 'objectValue', 'array', 'object');
        });
      });
    });
  });

  describe('When deep is true', () => {
    beforeEach(() => {
      target = create({
        arrayValue: [1, 2, 3],
        objectValue: { val1: 1, val2: '2', val3: true },
        method: () => false,
      });
    });

    it('should have child objects type checked', () => {
      expect(isTypeChecked(target.arrayValue)).toBe(true);
      expect(isTypeChecked(target.objectValue)).toBe(true);
    });

    it('should have methods type checked', () => {
      expect(isTypeChecked(target.method)).toBe(true);
    });

    describe('When accessing descendants', () => {
      beforeEach(() => {
        expect(target.objectValue.val1).toBe(1);
        expect(target.arrayValue[2]).toBe(3);
        target.arrayValue[2] = '5';
        target.objectValue.val1 = 'any string';
      });

      it('should report type violation', () => {
        expect(reporter).toHaveBeenCalledTimes(2);
        expect(reporter).toHaveBeenCalledWith(
          SET_PROPERTY,
          `arrayValue${INDEX}`,
          'number',
          'string',
        );
        expect(reporter).toHaveBeenCalledWith(
          SET_PROPERTY,
          'objectValue.val1',
          'number',
          'string',
        );
      });
    });
  });

  describe('When deep is false', () => {
    beforeEach(() => {
      target = create({
        arrayValue: [1, 2, 3],
        objectValue: { val1: 1, val2: '2', val3: true },
        method: () => false,
      }, { deep: false });
    });

    it('should have child objects type checked', () => {
      expect(isTypeChecked(target.arrayValue)).toBe(false);
      expect(isTypeChecked(target.objectValue)).toBe(false);
    });

    it('should have methods type checked', () => {
      expect(isTypeChecked(target.method)).toBe(true);
    });

    describe('When accessing descendants', () => {
      beforeEach(() => {
        target.arrayValue[2] = '5';
        target.objectValue.val1 = 'any string';
      });

      it('should ignore type violation', () => {
        expect(reporter).not.toHaveBeenCalled();
      });
    });
  });

  describe('When disabled', () => {
    beforeEach(() => {
      setEnabled(false);

      target = create({
        numberValue: 12,
      });

      afterEach(() => {
        setEnabled(true);
      });

      target.numberValue = 'abc';
      target.numberValue = false;
      target.numberValue = () => null;
    });

    it('should not be type checked', () => {
      expect(isTypeChecked(target)).toBe(false);
    });

    it('should not check types', () => {
      expect(reporter).not.toHaveBeenCalled();
    });
  });

  describe('When accessing types information', () => {
    beforeEach(() => {
      PrimitiveTypeChecker.collectTypesOnInit = true;
      target = create({
        numberValue: 12,
        stringValue: 'my string',
      });
    });

    it('should contain collection with primitive types', () => {
      expect(getTargetInfo(target)).toMatchSnapshot();
      expect(getTargetTypeCheckerConfig(target).types).toEqual({
        numberValue: 'number',
        stringValue: 'string',
      });
    });

    describe('When writing types information', () => {
      beforeEach(() => {
        const newTarget = create({
          booleanValue: true,
          arrayValue: [1, 2, 3],
          objectValue: { val1: 1, val2: '2', val3: true },
        });

        // init children types
        (() => null)(newTarget.arrayValue, newTarget.objectValue);

        // merge types
        setTargetInfo(target, getTargetInfo(newTarget));
      });

      it('should contain collection with primitive types', () => {
        expect(getTargetInfo(target)).toMatchSnapshot();
        expect(getTargetTypeCheckerConfig(target).types).toEqual({
          numberValue: 'number',
          stringValue: 'string',
          booleanValue: 'boolean',
          arrayValue: 'array',
          objectValue: 'object',
        });
      });
    });
  });

  describe('When accessing original object', () => {
    beforeEach(() => {
      PrimitiveTypeChecker.collectTypesOnInit = true;
      target = create({
        numberValue: 12,
        stringValue: 'my string',
      });
    });

    it('should be possible to access target object', () => {
      expect(getOriginalTarget(target)).toEqual({
        numberValue: 12,
        stringValue: 'my string',
        [INFO_KEY]: expect.any(Object),
      });
    });

    describe('When original object updated directly', () => {
      beforeEach(() => {
        getOriginalTarget(target).numberValue = '44';
        (() => null)(target.numberValue);
      });

      it('should report type error when accessing updated property', () => {
        expect(reporter).toHaveBeenCalledTimes(1);
        expect(reporter).toHaveBeenCalledWith(GET_PROPERTY, 'numberValue', 'number', 'string');
      });
    });

    describe('When reset original object', () => {
      it('should throw error', () => {
        expect(() => {
          target[TARGET_KEY] = { numberValue: 15 };
        }).toThrow();
      });

    });
  });
});
