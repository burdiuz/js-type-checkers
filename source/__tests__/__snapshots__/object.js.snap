// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Object When accessing types information When writing types information should contain collection with primitive types 1`] = `
Object {
  "checker": Object {
    "areArrayElementsOfSameType": true,
    "arguments": [Function],
    "collectTypesOnInit": true,
    "getIndexProperty": [Function],
    "getNamedProperty": [Function],
    "getProperty": [Function],
    "getTypeString": [Function],
    "ignorePrototypeValues": true,
    "init": [Function],
    "isIndexAccessTarget": [Function],
    "mergeConfigs": [Function],
    "registerIndexBasedClass": [Function],
    "replaceArgumentsTypeCheck": [Function],
    "replaceIndexedTypeCheck": [Function],
    "replacePropertyTypeCheck": [Function],
    "replaceReturnValueTypeCheck": [Function],
    "returnValue": [Function],
    "setIndexProperty": [Function],
    "setIndexValueType": [Function],
    "setIndexValueTypeBy": [Function],
    "setNamedProperty": [Function],
    "setProperty": [Function],
  },
  "children": Object {
    "arrayValue": Object {
      "checker": Object {
        "areArrayElementsOfSameType": true,
        "arguments": [Function],
        "collectTypesOnInit": true,
        "getIndexProperty": [Function],
        "getNamedProperty": [Function],
        "getProperty": [Function],
        "getTypeString": [Function],
        "ignorePrototypeValues": true,
        "init": [Function],
        "isIndexAccessTarget": [Function],
        "mergeConfigs": [Function],
        "registerIndexBasedClass": [Function],
        "replaceArgumentsTypeCheck": [Function],
        "replaceIndexedTypeCheck": [Function],
        "replacePropertyTypeCheck": [Function],
        "replaceReturnValueTypeCheck": [Function],
        "returnValue": [Function],
        "setIndexProperty": [Function],
        "setIndexValueType": [Function],
        "setIndexValueTypeBy": [Function],
        "setNamedProperty": [Function],
        "setProperty": [Function],
      },
      "children": Object {},
      "config": Object {
        "errorReporter": [MockFunction],
        "types": Object {
          "(Index)": "number",
        },
      },
      "deep": true,
      "names": Array [
        "arrayValue",
      ],
    },
    "objectValue": Object {
      "checker": Object {
        "areArrayElementsOfSameType": true,
        "arguments": [Function],
        "collectTypesOnInit": true,
        "getIndexProperty": [Function],
        "getNamedProperty": [Function],
        "getProperty": [Function],
        "getTypeString": [Function],
        "ignorePrototypeValues": true,
        "init": [Function],
        "isIndexAccessTarget": [Function],
        "mergeConfigs": [Function],
        "registerIndexBasedClass": [Function],
        "replaceArgumentsTypeCheck": [Function],
        "replaceIndexedTypeCheck": [Function],
        "replacePropertyTypeCheck": [Function],
        "replaceReturnValueTypeCheck": [Function],
        "returnValue": [Function],
        "setIndexProperty": [Function],
        "setIndexValueType": [Function],
        "setIndexValueTypeBy": [Function],
        "setNamedProperty": [Function],
        "setProperty": [Function],
      },
      "children": Object {},
      "config": Object {
        "errorReporter": [MockFunction],
        "types": Object {
          "val1": "number",
          "val2": "string",
          "val3": "boolean",
        },
      },
      "deep": true,
      "names": Array [
        "objectValue",
      ],
    },
  },
  "config": Object {
    "errorReporter": [MockFunction],
    "types": Object {
      "arrayValue": "array",
      "booleanValue": "boolean",
      "numberValue": "number",
      "objectValue": "object",
      "stringValue": "string",
    },
  },
  "deep": true,
  "names": Array [],
}
`;

exports[`Object When accessing types information should contain collection with primitive types 1`] = `
Object {
  "checker": Object {
    "areArrayElementsOfSameType": true,
    "arguments": [Function],
    "collectTypesOnInit": true,
    "getIndexProperty": [Function],
    "getNamedProperty": [Function],
    "getProperty": [Function],
    "getTypeString": [Function],
    "ignorePrototypeValues": true,
    "init": [Function],
    "isIndexAccessTarget": [Function],
    "mergeConfigs": [Function],
    "registerIndexBasedClass": [Function],
    "replaceArgumentsTypeCheck": [Function],
    "replaceIndexedTypeCheck": [Function],
    "replacePropertyTypeCheck": [Function],
    "replaceReturnValueTypeCheck": [Function],
    "returnValue": [Function],
    "setIndexProperty": [Function],
    "setIndexValueType": [Function],
    "setIndexValueTypeBy": [Function],
    "setNamedProperty": [Function],
    "setProperty": [Function],
  },
  "children": Object {},
  "config": Object {
    "errorReporter": [MockFunction],
    "types": Object {
      "numberValue": "number",
      "stringValue": "string",
    },
  },
  "deep": true,
  "names": Array [],
}
`;
